woocommerce/woocommerce-ios
# Core Data: Update transformable attribute type to NSArray

Closes: #14129 


## Description

We have always defined transformable attributes for our Core Data entities with Swift Array. This doesn't seem to have caused any issues in the past, but since Xcode 16, we got a lot of error logs in the console about this:

> fault: Declared Objective-C type "[Int64]" for attribute named relatedIDs is not valid

To avoid undefined behaviors due to this, this PR updates the transformable types declared in our Core Data model to NSArray instead of Swift Array types. The properties can be kept as Swift Array in our +CoreDataProperties extensions as Swift can cast NSArray to Swift Array types.

As mentioned in #14611, migration is unnecessary for these changes as they don't affect the architecture of the Core Data model, only the transformation step when data in the persistent container is transformed to types to work with the app.

Entities and attributes updated:

- `Coupon`: `emailRestrictions`, `excludedProductCategories`, `excludedProducts`, `productCategories`, `products`, `usedBy`.
- `PaymentGateway`: `features`
- `PaymentGatewayAccount`: `supportedCurrency`
- `Product`: `crossSellIDs`, `groupedProducts`, `relatedIDs`, `upsellIDs`, `variations`
- `ProductAttribute`: `options`
- `ProductBundleItem`: `allowedVariations`
- `ProductCompositeComponent`: `optionIDs`
- `ShippingLabel`: `productIDs` and `productNames`
- `Site`: `jetpackConnectionActivePlugins`
- `TaxRate`: `cities` and `postcodes`

## Steps to reproduce

Since there is no migration, writing migration tests doesn't make sense. The only way to confirm these changes work is to do manual testing. Let me know if you think otherwise.

Features to be tested: 
- Coupon
- Payments
- Variable product attributes
- Product bundle
- Composite products
- Shipping label
- Site picker
- Tax rate in order creation

To verify that the updates work correctly:
- Stay on the `trunk` branch and create data needed to verify the changes.
- Switch to this branch and confirm that all data are loaded successfully.
- Also keep an eye on Xcode console log and confirm errors about transformable attributes are gone.

## Testing information

Tested on simulator iPhone 16 Pro iOS 18.1 and confirmed that:

- The app builds and runs without crashing after the changes.
- No more Xcode console logs about transformable attributes.
- Smoked test the following features:
  - Coupons: confirmed that coupons with specified conditions for products/categories/emails are loaded correctly.
  - PaymentGateways are loaded successfully.
  - Product attributes for variable attributes are loaded successfully.
  - Shipping labels' product IDs and names are loaded successfully.
  - Sites' jetpackConnectionActivePlugins are loaded successfully.
  - Tax rates' cities and postcodes are loaded successfully.


## Screenshots

N/A

---

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:
